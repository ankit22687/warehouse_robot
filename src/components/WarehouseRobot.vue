<script setup lang="ts">
import { computed, reactive, ref } from "vue";
import { Gridcell, FormData } from "../interfaces/WarehouseRobotInterface";

const rows = 10;
const cols = 10;

const grid: Gridcell[][] = Array.from({ length: rows }, (_, rowIndex) =>
  Array.from({ length: cols }, (_, colIndex) => ({
    x: colIndex,
    y: rowIndex,
  }))
);

const robotPosition = ref<Gridcell>({ x: 0, y: 9 }); // Initial position at south-west corner

const errors = ref<FormData>({
  command: "",
});

const formData = reactive<FormData>({
  command: "",
});

const isCommandValid = computed(() => formData.command.trim() !== "");

const validateField = (field: keyof FormData) => {
  errors.value[field] = "";
  if (field === "command" && !isCommandValid.value) {
    errors.value.command = "Command is required.";
  }
};

// Function to get cell class based on position
const getCellClass = (row: number, col: number) => {
  return (
    "border border-gray-300 w-12 h-12 " +
    (robotPosition.value.x === col && robotPosition.value.y === row
      ? "bg-blue-500 text-white "
      : "bg-gray-100")
  );
};

const moveRobot = (commands: string[]) => {
  for (const command of commands) {
    switch (command) {
      case "N":
        if (robotPosition.value.y > 0) robotPosition.value.y--;
        break;
      case "S":
        if (robotPosition.value.y < 9) robotPosition.value.y++;
        break;
      case "W":
        if (robotPosition.value.x > 0) robotPosition.value.x--;
        break;
      case "E":
        if (robotPosition.value.x < 9) robotPosition.value.x++;
        break;
      default:
        // Invalid command
        break;
    }
  }
};

const handleSubmit = () => {
  errors.value.command = ""; // Clear previous errors
  validateField("command");

  // Convert commands to uppercase and split by whitespace
  const commandList = formData.command.trim().toUpperCase().split(" ");

  moveRobot(commandList);

  formData.command = "";
};
</script>

<template>
  <div class="container">
    <div class="flex flex-col justify-center items-center h-screen w-screen">
      <div class="flex flex-row my-10 w-96">
        <form @submit.prevent="handleSubmit">
          <div class="flex flex-row gap-2 w-full h-auto">
            <input
              type="text"
              v-model="formData.command"
              placeholder="Enter a command"
              name="command"
              id="command"
              class="rounded-lg"
            />

            <button
              class="text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 dark:bg-blue-600 dark:hover:bg-blue-700 focus:outline-none"
              type="submit"
            >
              Submit
            </button>
          </div>
          <div>
            <span
              class="error text-red-400"
              v-if="errors.command && !isCommandValid"
              >{{ errors.command }}</span
            >
          </div>
        </form>
      </div>
      <div class="grid gap-1">
        <div
          v-for="(row, rowIndex) in grid"
          :key="rowIndex"
          class="flex flex-row"
        >
          <div
            v-for="(cell, cellIndex) in row"
            :key="cellIndex"
            :class="getCellClass(cell.y, cell.x)"
            :data-x="cell.x"
            :data-y="cell.y"
          >
            <div
              v-if="robotPosition.x === cell.x && robotPosition.y === cell.y"
              class="flex items-center justify-center h-full w-full"
            >
              <i class="fa-solid fa-robot"></i>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>
