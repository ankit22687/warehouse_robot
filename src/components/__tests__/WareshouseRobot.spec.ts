import { describe, it, expect } from 'vitest';
import WarehouseRobot from '../WarehouseRobot.vue';
import { mount } from '@vue/test-utils';

describe('WarehouseRobot', () => {

    const wrapper = mount(WarehouseRobot);

    it('should render conrrectly', () => {
        expect(wrapper.html()).toMatchSnapshot();
    })

    it('should validate command', async () => {
        const form = wrapper.find('form');
        const input = wrapper.find('form').find('input[type="text"]');

        input.setValue('');
        await form.trigger('submit');

        expect(wrapper.find('.error').exists()).toBe(true);
        expect(wrapper.find('.error').text()).toContain('Command is required.');
    });

    it('should move the robot to original position', async () => {

        expect(wrapper.find('.bg-blue-500').exists()).toBe(true);

        const form = wrapper.find('form');
        const input = wrapper.find('form').find('input[type="text"]');

        input.setValue('N E S W');
        await form.trigger('submit');

        expect(wrapper.find('.bg-blue-500[data-x="0"][data-y="9"]').html()).toContain('<i class="fa-solid fa-robot"></i>');
    });

    it('should move the robot to middle of grid', async () => {

        expect(wrapper.find('.bg-blue-500').exists()).toBe(true);

        const form = wrapper.find('form');
        const input = wrapper.find('form').find('input[type="text"]');

        input.setValue('N E N E N E N E');
        await form.trigger('submit');

        expect(wrapper.find('.bg-blue-500[data-x="4"][data-y="5"]').html()).toContain('<i class="fa-solid fa-robot"></i>');
    });
});