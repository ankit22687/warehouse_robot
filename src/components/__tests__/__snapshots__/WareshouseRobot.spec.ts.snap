// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WarehouseRobot > should render conrrectly 1`] = `
"<div class="container">
  <div class="flex flex-col justify-center items-center h-screen w-screen">
    <div class="flex flex-row my-10 w-96">
      <form>
        <div class="flex flex-row gap-2 w-full h-auto"><input type="text" placeholder="Enter a command" name="command" id="command" class="rounded-lg"><button class="text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 dark:bg-blue-600 dark:hover:bg-blue-700 focus:outline-none" type="submit"> Submit </button></div>
        <div>
          <!--v-if-->
        </div>
      </form>
    </div>
    <div class="grid gap-1">
      <div class="flex flex-row">
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="0" data-y="0">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="1" data-y="0">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="2" data-y="0">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="3" data-y="0">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="4" data-y="0">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="5" data-y="0">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="6" data-y="0">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="7" data-y="0">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="8" data-y="0">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="9" data-y="0">
          <!--v-if-->
        </div>
      </div>
      <div class="flex flex-row">
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="0" data-y="1">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="1" data-y="1">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="2" data-y="1">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="3" data-y="1">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="4" data-y="1">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="5" data-y="1">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="6" data-y="1">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="7" data-y="1">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="8" data-y="1">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="9" data-y="1">
          <!--v-if-->
        </div>
      </div>
      <div class="flex flex-row">
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="0" data-y="2">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="1" data-y="2">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="2" data-y="2">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="3" data-y="2">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="4" data-y="2">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="5" data-y="2">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="6" data-y="2">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="7" data-y="2">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="8" data-y="2">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="9" data-y="2">
          <!--v-if-->
        </div>
      </div>
      <div class="flex flex-row">
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="0" data-y="3">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="1" data-y="3">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="2" data-y="3">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="3" data-y="3">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="4" data-y="3">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="5" data-y="3">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="6" data-y="3">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="7" data-y="3">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="8" data-y="3">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="9" data-y="3">
          <!--v-if-->
        </div>
      </div>
      <div class="flex flex-row">
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="0" data-y="4">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="1" data-y="4">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="2" data-y="4">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="3" data-y="4">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="4" data-y="4">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="5" data-y="4">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="6" data-y="4">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="7" data-y="4">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="8" data-y="4">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="9" data-y="4">
          <!--v-if-->
        </div>
      </div>
      <div class="flex flex-row">
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="0" data-y="5">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="1" data-y="5">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="2" data-y="5">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="3" data-y="5">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="4" data-y="5">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="5" data-y="5">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="6" data-y="5">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="7" data-y="5">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="8" data-y="5">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="9" data-y="5">
          <!--v-if-->
        </div>
      </div>
      <div class="flex flex-row">
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="0" data-y="6">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="1" data-y="6">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="2" data-y="6">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="3" data-y="6">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="4" data-y="6">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="5" data-y="6">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="6" data-y="6">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="7" data-y="6">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="8" data-y="6">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="9" data-y="6">
          <!--v-if-->
        </div>
      </div>
      <div class="flex flex-row">
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="0" data-y="7">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="1" data-y="7">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="2" data-y="7">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="3" data-y="7">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="4" data-y="7">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="5" data-y="7">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="6" data-y="7">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="7" data-y="7">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="8" data-y="7">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="9" data-y="7">
          <!--v-if-->
        </div>
      </div>
      <div class="flex flex-row">
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="0" data-y="8">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="1" data-y="8">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="2" data-y="8">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="3" data-y="8">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="4" data-y="8">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="5" data-y="8">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="6" data-y="8">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="7" data-y="8">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="8" data-y="8">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="9" data-y="8">
          <!--v-if-->
        </div>
      </div>
      <div class="flex flex-row">
        <div class="border border-gray-300 w-12 h-12 bg-blue-500 text-white" data-x="0" data-y="9">
          <div class="flex items-center justify-center h-full w-full"><i class="fa-solid fa-robot"></i></div>
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="1" data-y="9">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="2" data-y="9">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="3" data-y="9">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="4" data-y="9">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="5" data-y="9">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="6" data-y="9">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="7" data-y="9">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="8" data-y="9">
          <!--v-if-->
        </div>
        <div class="border border-gray-300 w-12 h-12 bg-gray-100" data-x="9" data-y="9">
          <!--v-if-->
        </div>
      </div>
    </div>
  </div>
</div>"
`;
